security.go

```diff
@@ -0,0 +1,50 @@
+package slack
+
+import (
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"net/http"
+)
+
+// SecretsVerifier contains the information needed to verify that the request comes from Slack
+type SecretsVerifier struct {
+	slackSig    string
+	timeStamp   string
+	requestBody string
+}
+
+// NewSecretsVerifierFromHeader returns a new SecretsVerifier object in exchange for an http.Header object
+func NewSecretsVerifierFromHeader(header http.Header) (SecretsVerifier, error) {
+	if header["X-Slack-Signature"][0] == "" || header["X-Slack-Request-Timestamp"][0] == "" {
+		return SecretsVerifier{}, errors.New("headers are empty, cannot create SecretsVerifier")
+	}
+
+	return SecretsVerifier{
+		slackSig:  header["X-Slack-Signature"][0],
+		timeStamp: header["X-Slack-Request-Timestamp"][0],
+	}, nil
+}
+
+func (v *SecretsVerifier) Write(body []byte) (n int, err error) {
+	v.requestBody = string(body)
+	return len(body), nil

```

Comment:
afraid this method is a bug, you may not get the entire content in a single invocation of this method. it should be:

```golang
func (v *SecretsVerifier) Write(body []byte) (n int, err error) {
     return v.hmac.Write(body)
}
```

Reply:
you need to concatenate "v0" and timestamp before hashing the resulting string, per Slack's instructions. will test to verify but I don't think this method will provide a valid signing secret
https://api.slack.com/docs/verifying-requests-from-slack

Reply:
ah I missed the prefix, just add it to the hash during initialization.

Reply:
```golang
hash := hmac.New(sha256.New, signingSecret)
hash.Write(prefix)
return SecretsVerifier{
	slackSig:  header["X-Slack-Signature"][0],
	timeStamp: header["X-Slack-Request-Timestamp"][0],
        hmac: hash,
}
```

Reply:
this still returns different hashes because you need to hash the prefix and the body together

Reply:
and just to make it [abundantly clear](https://play.golang.org/p/uiPH95wnZtK)

Reply:
just an FYI: most (all?) hashing algorithms are based on streams of data. so you can absolutely do partial writes for prefixes etc. and still end up with the same result. its the order of the content that gets hashed that matters.

Reply:
yeah sorry if my point was unclear, the wording was definitely bad. I get that you can write in a different order and still get the same result, my point was just that you aren't including it here so it will not return the right hash. also my point about not being able to do it all in the initializer was just that we would need to pass body in as an argument and that would complicate things so it doesn't seem like the best approach to me (the other option being calling the write function again on the request body back in the calling method) -- I'll make the changes and commit again to make it clearer, but thanks for being so thorough with your responses